File: single-number-iii.cpp
//Time - O(n)
//Space - O(1)
class Solution {
public:
    vector<int> singleNumber(vector<int>& nums) {
        int XOR = 0;
        
        for(int i=0;i<nums.size();i++){
            XOR = XOR^nums[i];
        }
        int mask = 1;
        while((XOR & mask )==0){
            mask = mask<<1;
        }
        int grp1 = 0,grp2 = 0;
        for(int i = 0;i<nums.size();i++){
            if(nums[i]&mask){
                grp1 = grp1^nums[i];
            }else{
                grp2 = grp2^nums[i];
            }
        }
        
        return {grp1,grp2};
    }
};

File: SingleNumberIII.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace BitManipulation
{
    class SingleNumberIII
    {
        /*
         * T.C: O(n) since interating through all the nums 
         * S.C: O(1)
         */
        public int[] SingleNumber(int[] nums)
        {
            if (nums == null || nums.Length == 0) return new int[] { };

            int xor = 0;

            foreach (int num in nums)
            {
                xor = xor ^ num;
            }

            int xor2 = 0;
            int temp = xor & (-xor);

            foreach (int num in nums)
            {
                if ((temp & num) != 0)
                {
                    xor2 = xor2 ^ num;
                }
            }

            return new int[] { xor2, xor ^ xor2 };
        }
    }
}
